src/bluesearch/database/article.py

```diff
@@ -184,17 +183,69 @@ def doi(self) -> str | None:
         """
         return None
 
+    @staticmethod
+    def get_uid_from_identifiers(identifiers: tuple[str | None, ...]) -> str:
+        """Generate a deterministic UID for a list of given paper identifiers.
+
+        Papers with the same values for the given identifiers get the same UID.
+
+        Missing values should have the value `None`, which is considered a value
+        by itself. Then, identifiers `(a, None)` and identifiers `(a, b)` have
+        two different UIDs.
+
+        Papers with all given identifiers unspecified have `None` as UID.
+
+        Parameters
+        ----------
+        identifiers
+            Values of the identifiers.
+
+        Returns
+        -------
+        str
+            A deterministic UID computed from the identifiers.
+
+        Raises
+        ------
+        ValueError
+            If all identifiers are `None`.
+        """
+        if all(x is None for x in identifiers):
+            raise ValueError(
+                f"Identifiers = {identifiers} are all `None`, UID cannot be computed."
+            )
+        else:
+            data = str(identifiers).encode()
+            hashed = hashlib.md5(data).hexdigest()  # nosec
+            return hashed
+
     @property
-    def uid(self) -> str | None:
-        """Generate unique ID of the article based on different identifiers.
+    def uid(self) -> str:
+        """Generate deterministic UID for an article.
+
+        The UID is usually created by hashing the identifiers of the article.
+        If no identifier is available, then the unique ID is computed by hashing
+        the whole content of the article.
 
         Returns
         -------
-        str or None
-            If at least one identifier exists, unique id is created. Otherwise,
-            the returned uid is None.
+        str
+            A deterministic UID.
         """
-        return generate_uid((self.pubmed_id, self.pmc_id, self.arxiv_id, self.doi))
+        identifiers = (self.pubmed_id, self.pmc_id, self.arxiv_id, self.doi)
+
+        # If no identifier is available, hash whole article content.
+        if all(x is None for x in identifiers):
+            m = hashlib.md5()  # nosec
+            m.update(self.title.encode())
+            m.update(str(list(self.authors)).encode())
+            m.update(str(list(self.abstract)).encode())
+            m.update(str(list(self.paragraphs)).encode())
```

Comment:
Yes, you are right that indeed changing the code responsible for parsing would cause a different `uid`.
At the same time, I don't know if there is any other solution?

Reply:
Maybe hashing the file itself (as it is something that is going to stay) ? But maybe it is too complicated. Let's keep it that way!

Reply:
Hashing the file would indeed be a valid option, but the problem is that, in general, instances of `ArticleParser` have no information about where the article contents are coming from.

A possible solution would be to make sure that the `__init__` of all `ArticleParser` subclasses takes in input the a `filename: Path | str` and stores it as an attribute `self.filename`. This would allow us to retrieve the original file and do whatever we want with it later on, including hashing it to compute the `uid`.

But:
1. Currently the `__init__` of most of the `ArticleParser` take in input the `xml` or `json` content read from file. So a bit of refactoring (maybe dedicated PR outside of this one?) would be required.
2. There is sometimes a `1-to-N` correspondence between files and articles, see e.g. `pubmed-xml-set`, so how would we deal with that?https://github.com/BlueBrain/Search/blob/2911ffc0e78fcebd3bbd95dae6225a3574d7144c/src/bluesearch/entrypoint/database/parse.py#L133-L136

----

Anyway, I would also be interested in hearing how @Stannislav and @jankrepl feel about this, we already had discussions about having the `__init__` consistent across `ArticleParser` subclasses and take in input the `filename`.

Reply:
## Discussed on 2022-02-01

Leave computation of `uid` as it is for now. But log a `logger.warning` to inform the user about this event.

Reply:
Added warning log in 66de14c.

Reply:
sure, lgtm